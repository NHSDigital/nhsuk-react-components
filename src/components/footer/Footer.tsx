import React, { HTMLProps } from 'react';
import classNames from 'classnames';
import { Container } from '../layout';

type FooterListProps = HTMLProps<HTMLOListElement>;

const FooterList: React.FC<FooterListProps> = ({ className, ...rest }) => (
  <ul className={classNames('nhsuk-footer__list', className)} {...rest} />
);

const FooterListItem: React.FC<HTMLProps<HTMLAnchorElement>> = ({ className, ...rest }) => (
  <li className="nhsuk-footer__list-item">
    <a className={classNames('nhsuk-footer__list-item-link', className)} {...rest} />
  </li>
);

const FooterCopyright: React.FC<HTMLProps<HTMLParagraphElement>> = ({ className, ...rest }) => (
  <p className={classNames('nhsuk-footer__copyright', className)} {...rest} />
);

interface FooterProps extends HTMLProps<HTMLDivElement> {
  visuallyHiddenText?: false | string;
}

interface Footer extends React.FC<FooterProps> {
  List: React.FC<FooterListProps>;
  ListItem: React.FC<HTMLProps<HTMLAnchorElement>>;
  Copyright: React.FC<HTMLProps<HTMLParagraphElement>>;
}

const Footer: Footer = ({ className, children, visuallyHiddenText, ...rest }) => (
  <footer {...rest}>
    <div className={classNames('nhsuk-footer', className)}>
      <Container>
        {visuallyHiddenText ? (
          <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
        ) : null}
        {children}
      </Container>
    </div>
  </footer>
);

Footer.defaultProps = {
  visuallyHiddenText: 'Support links',
};

Footer.List = FooterList;
Footer.ListItem = FooterListItem;
Footer.Copyright = FooterCopyright;

export default Footer;
