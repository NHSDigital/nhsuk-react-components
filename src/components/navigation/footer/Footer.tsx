/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { Children, FC, HTMLProps, cloneElement } from 'react';
import classNames from 'classnames';
import { Container } from '@components/layout';
import { childIsOfComponentType } from '@util/types/TypeGuards';

type FooterListProps = HTMLProps<HTMLOListElement> & { singleColumn?: boolean };

const FooterList: FC<FooterListProps> = ({
  className,
  children,
  singleColumn = false,
  ...rest
}) => {
  let newChildren = children;

  if (singleColumn) {
    newChildren = Children.map(newChildren, (child) =>
      childIsOfComponentType(child, FooterListItem) ? cloneElement(child, { singleColumn }) : child,
    );
  }

  return (
    <ul className={classNames('nhsuk-footer__list', className)} {...rest}>
      {newChildren}
    </ul>
  );
};

const FooterListItem: FC<HTMLProps<HTMLAnchorElement> & { singleColumn?: boolean }> = ({
  className,
  singleColumn = false,
  ...rest
}) => (
  <li
    className={classNames(
      'nhsuk-footer__list-item',
      singleColumn ? 'nhsuk-footer-default__list-item' : '',
    )}
  >
    <a className={classNames('nhsuk-footer__list-item-link', className)} {...rest} />
  </li>
);

const FooterCopyright: FC<HTMLProps<HTMLParagraphElement>> = ({ className, ...rest }) => (
  <p className={classNames('nhsuk-footer__copyright', className)} {...rest} />
);

interface FooterProps extends HTMLProps<HTMLDivElement> {
  visuallyHiddenText?: false | string;
}

interface Footer extends FC<FooterProps> {
  List: FC<FooterListProps>;
  ListItem: FC<HTMLProps<HTMLAnchorElement>>;
  Copyright: FC<HTMLProps<HTMLParagraphElement>>;
}

const Footer: Footer = ({ className, children, visuallyHiddenText = 'Support links', ...rest }) => {
  const footerCols = Children.toArray(children).filter((child) =>
    childIsOfComponentType(child, FooterList),
  );

  let newChildren = children;

  if (footerCols.length === 1) {
    newChildren = Children.map(children, (child) =>
      childIsOfComponentType(child, FooterList)
        ? cloneElement(child, { singleColumn: true })
        : child,
    );
  }

  return (
    <footer role="contentinfo" {...rest}>
      <div className={classNames('nhsuk-footer-container', className)}>
        <Container>
          {visuallyHiddenText ? (
            <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
          ) : null}
          <div className="nhsuk-footer">{newChildren}</div>
        </Container>
      </div>
    </footer>
  );
};

Footer.List = FooterList;
Footer.ListItem = FooterListItem;
Footer.Copyright = FooterCopyright;

export default Footer;
