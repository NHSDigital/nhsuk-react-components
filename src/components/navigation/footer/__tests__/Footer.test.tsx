import React from 'react';
import { render } from '@testing-library/react';
import Footer from '../';

jest.spyOn(console, 'warn').mockImplementation();

describe('Footer', () => {
  it('matches snapshot', () => {
    const { container } = render(<Footer />);

    expect(container).toMatchSnapshot('Footer');
  });
  it('has default visually hidden text', () => {
    const { container } = render(<Footer />);

    expect(container.querySelector('.nhsuk-u-visually-hidden')?.textContent).toBe('Support links');
  });

  it('has disabled visually hidden text', () => {
    const { container } = render(<Footer visuallyHiddenText={false} />);

    expect(container.querySelector('.nhsuk-u-visually-hidden')).toBeFalsy();
  });

  it('has custom visually hidden text', () => {
    const { container } = render(<Footer visuallyHiddenText="Custom" />);

    expect(container.querySelector('.nhsuk-u-visually-hidden')?.textContent).toBe('Custom');
  });

  it('Includes the single column class on ListItem when there is only one column', () => {
    const { container } = render(
      <Footer>
        <Footer.List>
          <Footer.ListItem id="test-listItem"></Footer.ListItem>
        </Footer.List>
      </Footer>,
    );

    expect(container.querySelector('#test-listItem')?.parentElement).toHaveClass(
      'nhsuk-footer-default__list-item',
    );
  });

  it('Does not include the single column class on ListItem when there is more than one column', () => {
    const { container } = render(
      <Footer>
        <Footer.List>
          <Footer.ListItem id="test-listItem"></Footer.ListItem>
        </Footer.List>
        <Footer.List>
          <Footer.ListItem id="test-listItem2"></Footer.ListItem>
        </Footer.List>
      </Footer>,
    );

    expect(container.querySelector('#test-listItem')?.parentElement).not.toHaveClass(
      'nhsuk-footer-default__list-item',
    );
    expect(container.querySelector('#test-listItem2')?.parentElement).not.toHaveClass(
      'nhsuk-footer-default__list-item',
    );
  });

  describe('Footer.List', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('matches snapshot', () => {
      const { container } = render(<Footer.List />);

      expect(container).toMatchSnapshot('Footer.List');
    });
  });

  describe('Footer.ListItem', () => {
    it('matches snapshot', () => {
      const { container } = render(<Footer.ListItem />);

      expect(container).toMatchSnapshot('Footer.ListItem');
    });
  });

  describe('Footer.Copyright', () => {
    it('matches snapshot', () => {
      const { container } = render(<Footer.Copyright />);

      expect(container).toMatchSnapshot('Footer.Copyright');
    });
  });
});
