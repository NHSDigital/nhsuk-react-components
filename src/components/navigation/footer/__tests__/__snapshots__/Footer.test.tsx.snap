// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Footer.Copyright matches snapshot: Footer.Copyright 1`] = `
<p
  className="nhsuk-footer__copyright"
/>
`;

exports[`Footer Footer.List matches snapshot: Footer.List 1`] = `
<ul
  className="nhsuk-footer__list"
/>
`;

exports[`Footer Footer.ListItem matches snapshot: Footer.ListItem 1`] = `
<li
  className="nhsuk-footer__list-item"
>
  <a
    className="nhsuk-footer__list-item-link"
  />
</li>
`;

exports[`Footer matches snapshot: Footer 1`] = `
<footer>
  <div
    className="nhsuk-footer"
  >
    <Container>
      <h2
        className="nhsuk-u-visually-hidden"
      >
        Support links
      </h2>
    </Container>
  </div>
</footer>
`;
